File: app/internal/publisher/template/cpp/license_engine.cpp
#include "license_engine.h"
#include "client.h"
#include "hardware.h"
#include "cache.h"
#include "config.h"
#include <fstream>
#include <filesystem>
#include <stdexcept>

namespace fs = std::filesystem;

LicenseStatus LicenseStatus::from_dict(const std::map<std::string, std::string>& data) {
    LicenseStatus s;
    s.valid = data.count("valid") && data.at("valid") == "true";
    auto it = data.find("status");
    s.status = (it != data.end()) ? it->second : "unlicensed";
    it = data.find("expires_at");
    if (it != data.end()) s.expires_at = it->second;
    it = data.find("days_remaining");
    if (it != data.end()) s.days_remaining = std::stoi(it->second);
    it = data.find("plan");
    if (it != data.end()) s.plan = it->second;
    it = data.find("hardware_id");
    if (it != data.end()) s.hardware_id = it->second;
    it = data.find("message");
    if (it != data.end()) s.message = it->second;
    it = data.find("license_key");
    if (it != data.end()) s.license_key = it->second;
    it = data.find("trial_active");
    s.trial_active = (it != data.end()) ? (it->second == "true") : (s.status == "trial");
    return s;
}

std::map<std::string, std::string> LicenseStatus::to_dict() const {
    std::map<std::string, std::string> d;
    d["valid"] = valid ? "true" : "false";
    d["status"] = status;
    if (!expires_at.empty()) d["expires_at"] = expires_at;
    d["days_remaining"] = std::to_string(days_remaining);
    if (!plan.empty()) d["plan"] = plan;
    if (!hardware_id.empty()) d["hardware_id"] = hardware_id;
    if (!message.empty()) d["message"] = message;
    if (!license_key.empty()) d["license_key"] = license_key;
    d["trial_active"] = trial_active ? "true" : "false";
    return d;
}

LicenseEngine::LicenseEngine(const std::string& config_path) {
    config = load_config(config_path.empty() ? find_config_path("") : config_path);
    hardware_ = std::make_shared<HardwareDetector>();
    cache_ = std::make_shared<CacheManager>(config);
    client_ = std::make_shared<ApiClient>(config, hardware_, cache_);
}

LicenseEngine::~LicenseEngine() {}

std::string LicenseEngine::find_config_path(const std::string& config_path) {
    if (!config_path.empty() && fs::exists(config_path)) return config_path;
    std::vector<std::string> candidates = {
        "config/api-config.json",
        "../config/api-config.json",
    };
    for (const auto& p : candidates) {
        if (fs::exists(p)) return fs::absolute(p).string();
    }
    return config_path;
}

std::map<std::string, std::string> LicenseEngine::load_config(const std::string& config_path) {
    std::map<std::string, std::string> cfg;
    if (!fs::exists(config_path)) {
        cfg["product_json"] = "{}";
        cfg["api_json"] = "{}";
        cfg["offline_json"] = "{}";
        cfg["branding_json"] = "{}";
        cfg["trial_json"] = "{}";
        cfg["license_json"] = "{}";
        cfg["hardware_json"] = "{}";
        cfg["security_json"] = "{}";
        cfg["features_json"] = "{}";
        return cfg;
    }
    std::ifstream f(config_path);
    std::string content((std::istreambuf_iterator<char>(f)),
                         std::istreambuf_iterator<char>());
    f.close();
    auto root = json_parse(content);
    for (const auto& [k, v] : root) {
        cfg[k + "_json"] = v;
    }
    auto prod = json_parse(root["product"]);
    for (const auto& [k, v] : prod) cfg["product_" + k] = v;
    auto api = json_parse(root["api"]);
    for (const auto& [k, v] : api) cfg["api_" + k] = v;
    auto lic = json_parse(root["license"]);
    for (const auto& [k, v] : lic) cfg["license_" + k] = v;
    auto trial = json_parse(root["trial"]);
    for (const auto& [k, v] : trial) cfg["trial_" + k] = v;
    return cfg;
}

LicenseStatus LicenseEngine::initialize() {
    if (cache_->is_valid()) {
        auto cached = cache_->get_license_status();
        if (cached) {
            status_ = LicenseStatus::from_dict(*cached);
            has_status_ = true;
            return status_;
        }
    }
    try {
        std::string hardware_id = hardware_->get_fingerprint();
        auto trial_response = client_->get_trial_status(hardware_id);
        std::string data_json = trial_response["data_json"];
        auto trial_data = json_parse(data_json);
        if (trial_data["has_trial"] == "true") {
            std::string status_str = trial_data["status"];
            if (status_str.empty()) status_str = "trial";
            status_.valid = (status_str == "active");
            status_.status = status_str;
            status_.expires_at = trial_data["expiry_date"];
            auto dl = trial_data.find("days_left");
            status_.days_remaining = (dl != trial_data.end() && !dl->second.empty())
                ? std::stoi(dl->second) : 0;
            status_.plan = trial_data["plan"];
            status_.hardware_id = hardware_id;
            status_.message = "Trial is " + status_str;
            status_.trial_active = true;
            has_status_ = true;
            if (status_.valid) cache_->set_license_status(status_.to_dict());
            return status_;
        }
        status_.valid = false;
        status_.status = "unlicensed";
        status_.hardware_id = hardware_id;
        status_.message = "No license or trial found";
        has_status_ = true;
        return status_;
    } catch (const std::exception& e) {
        auto cached = cache_->get_license_status();
        if (cached) {
            status_ = LicenseStatus::from_dict(*cached);
            has_status_ = true;
            return status_;
        }
        status_.valid = false;
        status_.status = "error";
        status_.message = std::string("Unexpected error: ") + e.what();
        has_status_ = true;
        return status_;
    }
}

std::string LicenseEngine::get_hardware_id() {
    return hardware_->get_fingerprint();
}

LicenseStatus* LicenseEngine::get_status() {
    return has_status_ ? &status_ : nullptr;
}

std::string LicenseEngine::get_license_key() {
    return license_key_;
}

bool LicenseEngine::has_license_key() {
    return !license_key_.empty();
}

std::map<std::string, std::string> LicenseEngine::validate(const std::string& license_key) {
    std::string key = license_key.empty() ? license_key_ : license_key;
    if (key.empty()) {
        std::map<std::string, std::string> err;
        err["success"] = "false";
        err["message"] = "License key unavailable. Please activate first.";
        return err;
    }
    std::string hardware_id = hardware_->get_fingerprint();
    auto result = client_->validate_license(key, hardware_id);
    auto data = json_parse(result["data_json"]);
    if (data["valid"] == "true") {
        if (!data["license_key"].empty()) license_key_ = data["license_key"];
        initialize();
    }
    return result;
}

std::map<std::string, std::string> LicenseEngine::activate(const std::string& license_key) {
    auto result = client_->activate_license(license_key);
    if (result["success"] == "true") {
        license_key_ = license_key;
        initialize();
    }
    return result;
}

std::map<std::string, std::string> LicenseEngine::start_trial(
    const std::string& email, const std::string& customer_name,
    const std::map<std::string, std::string>& customer_data) {
    auto result = client_->start_trial(email, customer_name, customer_data);
    if (result["success"] == "true") initialize();
    return result;
}

std::map<std::string, std::string> LicenseEngine::convert_trial(
    const std::string& plan, const std::string& customer_name,
    const std::string& customer_email) {
    auto status = initialize();
    if (status.status != "trial") {
        std::map<std::string, std::string> err;
        err["success"] = "false";
        err["message"] = "No active trial to convert.";
        return err;
    }
    std::string hardware_id = hardware_->get_fingerprint();
    auto result = client_->convert_trial(hardware_id, plan, customer_name, customer_email);
    if (result["success"] == "true") {
        if (result.count("license_key")) license_key_ = result["license_key"];
        initialize();
    }
    return result;
}

std::map<std::string, std::string> LicenseEngine::renew(int extra_days) {
    if (license_key_.empty()) {
        std::map<std::string, std::string> err;
        err["success"] = "false";
        err["message"] = "License key unavailable. Please activate first.";
        return err;
    }
    auto result = client_->renew_license(license_key_, extra_days);
    if (result["success"] == "true") initialize();
    return result;
}

std::map<std::string, std::string> LicenseEngine::deactivate(const std::string& license_key) {
    std::string key = license_key.empty() ? license_key_ : license_key;
    if (key.empty()) {
        std::map<std::string, std::string> err;
        err["success"] = "false";
        err["message"] = "License key unavailable. Please provide a key.";
        return err;
    }
    auto result = client_->deactivate_license(key);
    if (result["success"] == "true") {
        cache_->invalidate_license_status();
        has_status_ = false;
        if (license_key.empty()) license_key_.clear();
    }
    return result;
}

std::map<std::string, std::string> LicenseEngine::replace_hardware() {
    if (license_key_.empty()) {
        std::map<std::string, std::string> err;
        err["success"] = "false";
        err["message"] = "License key unavailable. Please activate first.";
        return err;
    }
    std::string new_hardware_id = hardware_->get_fingerprint();
    std::string old_hardware_id;
    if (has_status_ && !status_.hardware_id.empty())
        old_hardware_id = status_.hardware_id;
    if (old_hardware_id.empty()) {
        auto cached = cache_->get_license_status();
        if (cached) old_hardware_id = (*cached)["hardware_id"];
    }
    if (old_hardware_id.empty()) {
        std::map<std::string, std::string> err;
        err["success"] = "false";
        err["message"] = "Current hardware_id unavailable. Cannot replace device.";
        return err;
    }
    if (old_hardware_id == new_hardware_id) {
        std::map<std::string, std::string> err;
        err["success"] = "false";
        err["message"] = "Old and new hardware IDs are identical.";
        return err;
    }
    auto result = client_->replace_device(license_key_, new_hardware_id, old_hardware_id);
    if (result["success"] == "true") {
        cache_->invalidate_license_status();
        has_status_ = false;
        initialize();
    }
    return result;
}

std::map<std::string, std::string> LicenseEngine::bind_device(
    const std::string& license_key, const std::string& device_name) {
    std::string key = license_key.empty() ? license_key_ : license_key;
    if (key.empty()) {
        std::map<std::string, std::string> err;
        err["success"] = "false";
        err["message"] = "License key unavailable.";
        return err;
    }
    auto result = client_->bind_device(key, "", device_name);
    if (result["success"] == "true") initialize();
    return result;
}

std::map<std::string, std::string> LicenseEngine::get_plans() {
    return client_->get_products();
}

File: app/internal/publisher/template/dotnet/LicenseEngine.cs
using System;
using System.IO;
using System.Text.Json;
using System.Threading.Tasks;

namespace WebsmithSDK;

public class LicenseEngine
{
    private readonly ApiClient _client;
    private readonly CacheManager _cache;
    private readonly HardwareDetector _hardware;
    private readonly string _hardwareId;
    private string? _licenseKey;
    private JsonDocument? _status;

    public LicenseEngine() : this("config/api-config.json") { }

    public LicenseEngine(string configPath)
    {
        var configText = File.ReadAllText(configPath);
        var config = JsonDocument.Parse(configText);
        _hardware = new HardwareDetector();
        _cache = new CacheManager(config);
        _client = new ApiClient(config, _hardware, _cache);
        _hardwareId = _hardware.GetFingerprint();
    }

    public ApiClient GetClient() => _client;

    public LicenseStatus Initialize()
    {
        if (_cache.IsValid())
        {
            var cached = _cache.GetLicenseStatus();
            if (cached != null)
            {
                try
                {
                    _status = JsonDocument.Parse(cached);
                    return LicenseStatus.FromJson(_status.RootElement);
                }
                catch { }
            }
        }
        try
        {
            var trialResponse = _client.GetTrialStatus(_hardwareId).Result;
            JsonElement trialData;
            if (trialResponse.RootElement.TryGetProperty("data", out var d))
                trialData = d;
            else
                trialData = trialResponse.RootElement;

            if (trialData.TryGetProperty("has_trial", out var hasTrial) && hasTrial.GetBoolean())
            {
                var statusStr = trialData.TryGetProperty("status", out var st) ? st.GetString() ?? "trial" : "trial";
                var ls = new LicenseStatus(
                    statusStr == "active",
                    statusStr,
                    trialData.TryGetProperty("expiry_date", out var ed) ? ed.GetString() : null,
                    trialData.TryGetProperty("days_left", out var dl) ? dl.GetInt32() : 0,
                    trialData.TryGetProperty("plan", out var p) ? p.GetString() : null,
                    _hardwareId,
                    $"Trial is {statusStr}",
                    null,
                    statusStr == "trial"
                );
                if (ls.Valid)
                    _cache.SetLicenseStatus(ls.ToJson().RootElement.GetRawText());
                _status = ls.ToJson();
                return ls;
            }

            var unlicensed = new LicenseStatus(false, "unlicensed", _hardwareId,
                "No license or trial found");
            _status = unlicensed.ToJson();
            return unlicensed;
        }
        catch (Exception e)
        {
            var cached = _cache.GetLicenseStatus();
            if (cached != null)
            {
                try
                {
                    _status = JsonDocument.Parse(cached);
                    return LicenseStatus.FromJson(_status.RootElement);
                }
                catch { }
            }
            var err = new LicenseStatus(false, "error", null,
                $"Unexpected error: {e.Message}");
            _status = err.ToJson();
            return err;
        }
    }

    public string GetHardwareId() => _hardwareId;
    public string? GetLicenseKey() => _licenseKey;
    public bool HasLicenseKey() => !string.IsNullOrEmpty(_licenseKey);

    public bool IsValid()
    {
        if (_status == null) return false;
        var root = _status.RootElement;
        if (!root.TryGetProperty("status", out var status) || status.GetString() != "active")
            return false;
        if (root.TryGetProperty("expires_at", out var expiresAt) && expiresAt.ValueKind == JsonValueKind.String)
        {
            if (DateTime.TryParse(expiresAt.GetString(), out var exp) && exp < DateTime.UtcNow)
                return false;
        }
        return true;
    }

    public LicenseStatus? GetStatus()
    {
        if (_status == null) return null;
        return LicenseStatus.FromJson(_status.RootElement);
    }

    public async Task<JsonDocument> Validate(string? licenseKey = null)
    {
        var key = licenseKey ?? _licenseKey;
        if (string.IsNullOrEmpty(key))
            throw new InvalidOperationException("License key unavailable. Please activate first.");

        var result = await _client.ValidateLicense(key, _hardwareId);
        JsonElement data;
        if (result.RootElement.TryGetProperty("data", out var d))
            data = d;
        else
            data = result.RootElement;

        if (data.TryGetProperty("valid", out var v) && v.GetBoolean())
        {
            if (data.TryGetProperty("license_key", out var lk))
                _licenseKey = lk.GetString();
            Initialize();
        }
        return result;
    }

    public async Task<JsonDocument> Activate(string licenseKey)
    {
        var result = await _client.ActivateLicense(licenseKey, _hardwareId);
        if (result.RootElement.TryGetProperty("success", out var s) && s.GetBoolean())
        {
            _licenseKey = licenseKey;
            Initialize();
        }
        return result;
    }

    public async Task<JsonDocument> StartTrial(string email, string customerName, Dictionary<string, object>? customerData = null)
    {
        var result = await _client.StartTrial(email, customerName, _hardwareId, customerData);
        if (result.RootElement.TryGetProperty("success", out var s) && s.GetBoolean())
            Initialize();
        return result;
    }

    public async Task<JsonDocument> ConvertTrial(string? plan, string customerName, string customerEmail)
    {
        var ls = GetStatus();
        if (ls == null || ls.Status != "trial")
            throw new InvalidOperationException("No active trial to convert.");

        var result = await _client.ConvertTrial(_hardwareId, plan, customerName, customerEmail);
        if (result.RootElement.TryGetProperty("success", out var s) && s.GetBoolean())
        {
            if (result.RootElement.TryGetProperty("license_key", out var lk))
                _licenseKey = lk.GetString();
            Initialize();
        }
        return result;
    }

    public async Task<JsonDocument> Renew(int? extraDays = null)
    {
        if (string.IsNullOrEmpty(_licenseKey))
            throw new InvalidOperationException("License key unavailable. Please activate first.");

        var result = await _client.RenewLicense(_licenseKey, extraDays);
        if (result.RootElement.TryGetProperty("success", out var s) && s.GetBoolean())
            Initialize();
        return result;
    }

    public async Task<JsonDocument> Deactivate(string? licenseKey = null)
    {
        var key = licenseKey ?? _licenseKey;
        if (string.IsNullOrEmpty(key))
            throw new InvalidOperationException("License key unavailable. Please provide a key.");

        var result = await _client.DeactivateLicense(key, _hardwareId);
        if (result.RootElement.TryGetProperty("success", out var s) && s.GetBoolean())
        {
            _cache.InvalidateLicenseStatus();
            _status = null;
            if (licenseKey == null)
                _licenseKey = null;
        }
        return result;
    }

    public async Task<JsonDocument> ReplaceHardware()
    {
        if (string.IsNullOrEmpty(_licenseKey))
            throw new InvalidOperationException("License key unavailable. Please activate first.");

        var newHardwareId = _hardware.GetFingerprint();
        string? oldHardwareId = null;

        if (_status != null && _status.RootElement.TryGetProperty("hardware_id", out var hw))
            oldHardwareId = hw.GetString();

        if (oldHardwareId == null)
        {
            var cached = _cache.GetLicenseStatus();
            if (cached != null)
            {
                try
                {
                    var doc = JsonDocument.Parse(cached);
                    if (doc.RootElement.TryGetProperty("hardware_id", out var chw))
                        oldHardwareId = chw.GetString();
                }
                catch { }
            }
        }

        if (oldHardwareId == null)
            throw new InvalidOperationException("Current hardware_id unavailable. Cannot replace device.");

        if (oldHardwareId == newHardwareId)
            return JsonDocument.Parse("{\"success\":false,\"message\":\"Old and new hardware IDs are identical.\"}");

        var result = await _client.ReplaceDevice(_licenseKey, oldHardwareId, newHardwareId);
        if (result.RootElement.TryGetProperty("success", out var s) && s.GetBoolean())
        {
            _cache.InvalidateLicenseStatus();
            _status = null;
            Initialize();
        }
        return result;
    }

    public async Task<JsonDocument> BindDevice(string? licenseKey = null, string? deviceName = null)
    {
        var key = licenseKey ?? _licenseKey;
        if (string.IsNullOrEmpty(key))
            throw new InvalidOperationException("License key unavailable.");

        var result = await _client.BindDevice(key, _hardwareId, deviceName);
        if (result.RootElement.TryGetProperty("success", out var s) && s.GetBoolean())
            Initialize();
        return result;
    }
}

public class LicenseStatus
{
    public bool Valid { get; }
    public string Status { get; }
    public string? ExpiresAt { get; }
    public int DaysRemaining { get; }
    public string? Plan { get; }
    public string? HardwareId { get; }
    public string? Message { get; }
    public string? LicenseKey { get; }
    public bool TrialActive { get; }

    public LicenseStatus(bool valid, string status, string? expiresAt, int daysRemaining,
        string? plan, string? hardwareId, string? message, string? licenseKey, bool trialActive)
    {
        Valid = valid;
        Status = status;
        ExpiresAt = expiresAt;
        DaysRemaining = daysRemaining;
        Plan = plan;
        HardwareId = hardwareId;
        Message = message;
        LicenseKey = licenseKey;
        TrialActive = trialActive;
    }

    public LicenseStatus(bool valid, string status, string? hardwareId, string? message)
        : this(valid, status, null, 0, null, hardwareId, message, null, status == "trial") { }

    public static LicenseStatus FromJson(JsonElement data)
    {
        var status = data.TryGetProperty("status", out var s) ? s.GetString() ?? "unlicensed" : "unlicensed";
        return new LicenseStatus(
            data.TryGetProperty("valid", out var v) && v.GetBoolean(),
            status,
            data.TryGetProperty("expires_at", out var ea) && ea.ValueKind == JsonValueKind.String ? ea.GetString() : null,
            data.TryGetProperty("days_remaining", out var dr) ? dr.GetInt32() : 0,
            data.TryGetProperty("plan", out var p) && p.ValueKind == JsonValueKind.String ? p.GetString() : null,
            data.TryGetProperty("hardware_id", out var hw) && hw.ValueKind == JsonValueKind.String ? hw.GetString() : null,
            data.TryGetProperty("message", out var m) && m.ValueKind == JsonValueKind.String ? m.GetString() : null,
            data.TryGetProperty("license_key", out var lk) && lk.ValueKind == JsonValueKind.String ? lk.GetString() : null,
            data.TryGetProperty("trial_active", out var ta) ? ta.GetBoolean() : status == "trial"
        );
    }

    public JsonDocument ToJson()
    {
        var obj = new Dictionary<string, object?>
        {
            ["valid"] = Valid,
            ["status"] = Status,
            ["expires_at"] = ExpiresAt,
            ["days_remaining"] = DaysRemaining,
            ["plan"] = Plan,
            ["hardware_id"] = HardwareId,
            ["message"] = Message,
            ["license_key"] = LicenseKey,
            ["trial_active"] = TrialActive
        };
        var json = JsonSerializer.Serialize(obj);
        return JsonDocument.Parse(json);
    }
}
